File: src/btree/c_arena/main_arena.c
// SPDX-FileCopyrightText: 2026 TSUKUMO Akito <tsukumoakito99@duck.com>
// SPDX-License-Identifier: MIT

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  struct Node *left;
  struct Node *right;
} Node;

typedef struct {
  char *buffer;
  size_t capacity;
  size_t offset;
} Arena;

void *arena_alloc(Arena *arena, size_t size) {
  if (arena->offset + size > arena->capacity)
    return NULL;
  void *ptr = &arena->buffer[arena->offset];
  arena->offset += size;
  return ptr;
}

Node *createTree(Arena *arena, int depth) {
  Node *node = (Node *)arena_alloc(arena, sizeof(Node));
  if (depth > 0) {
    node->left = createTree(arena, depth - 1);
    node->right = createTree(arena, depth - 1);
  } else {
    node->left = NULL;
    node->right = NULL;
  }
  return node;
}

uint64_t countNodes(Node *node) {
  if (!node)
    return 0;
  return 1 + countNodes(node->left) + countNodes(node->right);
}

int main(int argc, char *argv[]) {
  int depth = 20;
  if (argc > 1)
    depth = atoi(argv[1]);

  size_t total_nodes = (1ULL << (depth + 1)) - 1;
  size_t required_size = total_nodes * sizeof(Node);

  Arena arena;
  arena.capacity = required_size;
  arena.buffer = (char *)malloc(arena.capacity);
  arena.offset = 0;

  if (!arena.buffer) {
    fprintf(stderr, "Failed to allocate arena of size %zu\n", required_size);
    return 1;
  }

  Node *root = createTree(&arena, depth);
  uint64_t count = countNodes(root);

  printf("Checksum: %llu\n", (unsigned long long)count);
  fprintf(stderr, "Depth: %d\n", depth);

  free(arena.buffer);

  return 0;
}

File: src/btree/rust_arena/main_arena.rs
// SPDX-FileCopyrightText: 2026 TSUKUMO Akito <tsukumoakito99@duck.com>
// SPDX-License-Identifier: MIT

use std::env;
use std::io::{self, Write};

struct Node {
    left: Option<usize>,
    right: Option<usize>,
}

struct Arena {
    nodes: Vec<Node>,
}

impl Arena {
    fn new(capacity: usize) -> Self {
        Arena {
            nodes: Vec::with_capacity(capacity),
        }
    }

    fn alloc_node(&mut self) -> usize {
        let idx = self.nodes.len();
        self.nodes.push(Node {
            left: None,
            right: None,
        });
        idx
    }

    fn create_tree(&mut self, depth: i32) -> usize {
        let node_idx = self.alloc_node();
        if depth > 0 {
            let left = self.create_tree(depth - 1);
            let right = self.create_tree(depth - 1);
            self.nodes[node_idx].left = Some(left);
            self.nodes[node_idx].right = Some(right);
        }
        node_idx
    }

    fn count_nodes(&self, node_idx: usize) -> u64 {
        let mut count = 1;
        let node = &self.nodes[node_idx];
        if let Some(l) = node.left {
            count += self.count_nodes(l);
        }
        if let Some(r) = node.right {
            count += self.count_nodes(r);
        }
        count
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let depth: i32 = if args.len() > 1 {
        args[1].parse().unwrap_or(20)
    } else {
        20
    };

    let total_nodes = (1usize << (depth + 1)) - 1;

    let mut arena = Arena::new(total_nodes);

    let root_idx = arena.create_tree(depth);
    let count = arena.count_nodes(root_idx);

    println!("Checksum: {}", count);
    let mut stderr = io::stderr();
    let _ = writeln!(stderr, "Depth: {}", depth);

}
